blog/2022-11-21-golang-memory-dump.md

```diff
@@ -0,0 +1,187 @@
+         0     0% 92.43%    25.41GB 91.78%  github.com/polarsignals/pprof-example-app-go/fib.Fibonacci
+         0     0% 92.43%    25.41GB 91.78%  math/big.(*Int).Add
+         0     0% 92.43%    25.41GB 91.78%  math/big.nat.add
+    2.02GB  7.30% 99.73%     2.02GB  7.30%  main.allocMem
+    0.02GB 0.073% 99.80%     0.25GB  0.91%  fmt.Sprintln
+         0     0% 99.80%     0.25GB  0.91%  log.Println
+         0     0% 99.80%     0.23GB  0.84%  fmt.(*pp).doPrintln
+         0     0% 99.80%     0.23GB  0.84%  fmt.(*pp).handleMethods
+```
+
+`top 10 -flat` will show you the resource consumption it takes to call a function excluding all function it calls
+
+```pprof
+Active filters:
+   ignore=flat
+Showing nodes accounting for 27.66GB, 99.90% of 27.69GB total
+Dropped 10 nodes (cum <= 0.14GB)
+Showing top 10 nodes out of 17
+      flat  flat%   sum%        cum   cum%
+   25.59GB 92.43% 92.43%    25.59GB 92.43%  math/big.nat.make (inline)
+    2.02GB  7.30% 99.73%     2.02GB  7.30%  main.allocMem
+    0.03GB   0.1% 99.83%     0.21GB  0.76%  math/big.nat.itoa
+    0.02GB 0.073% 99.90%     0.25GB  0.91%  fmt.Sprintln
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).doPrintln
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).handleMethods
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).printArg
+         0     0% 99.90%    25.41GB 91.78%  github.com/polarsignals/pprof-example-app-go/fib.Fibonacci
+         0     0% 99.90%     0.25GB  0.91%  log.Println
+         0     0% 99.90%    25.66GB 92.69%  main.calculateFib
+```
+
+By looking at the output we can see that it's `math/big.nat.make` that takes the most resources.
+
+Just for fun I also generated a flame graph using pprof by port-forwarding to the application and running
+
+```shell
+go tool pprof -http=: http://localhost:8080/debug/pprof/allocs
+```
+
+<img alt="pprof flame" src={useBaseUrl("img/assets/blog/pprof_flame.png")} />
+
+## Cleanup
+
+To cleanup the resources we created run:
+
+```shell
+kubectl delete -f https://raw.githubusercontent.com/polarsignals/pprof-example-app-go/main/manifests/deployment.yaml
+kubectl delete svc pprof-example-app-go
+kubectl delete pod curl
+```
+
+## Conclusion
+
+The `kubectl debug` command is extremely useful when you want to debug your application and you
+don't have access to a shell or the tools that you need in your normal container.
+
+It saves us from having to install unneeded applications in our container which lowers the amount of potential CVE:s and the time it takes to start your container by lower container size.
+
+Kubectl debug isn't perfect and it won't work for all your uses cases especially since you can't
+attach volumes but it's a great start.
+
+If you really don't want to use a tool like pprof inside a container you could of course write some small application that performs a request to an endpoint and streams the output to an object store or similar without ever storing the data on disk.
```

Comment:
Not sure this para contributes much?

Reply:
I guessed I wanted to share another possible solution. But you have a good point.
I will remove it.